### PR TITLE
906 http proxy service

### PR DESCRIPTION
**Pull request by jonahx**
History is clean: Use "Commits" tab to review.

Also note: None of the CC errors are new code.  They were all introduced in the final comment when I moved stuff around to fix a circular dep problem.  And almost all of them will be fixed when we delete the old code.

### CODE REVIEW

File: internal/proxyservice/proxy_service.go [hunk 1]
```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
...
+			// we won't be unit testing this.  Instead, we'll make the change
+			// in the TODO_ above and this code will be deleted then.
+			log.Fatalln("Nil configuration is not allowed!")
+		}
+		rawConfigs = append(rawConfigs, *cfg)
+	}
+
+	// Separate HTTP and TCP configs
+	var httpConfigs []v2.Service
+	var	tcpConfigs []v2.Service
```
Comment (sgnn7) [inline internal/proxyservice/proxy_service.go]: nit: Spacing

File: internal/proxyservice/proxy_service.go [hunk 2]
```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
...
+
+	// TODO: Add proper validation here of the type.  This should moved into
+	//   IsHTTPPlugin, whose API will likely change to returning a type or an
+	//   error
+	for _, cfg := range rawConfigs {
+		if plugin.IsHTTPPlugin(availPlugins, cfg.Connector) {
+			httpConfigs = append(httpConfigs, cfg)
```
Comment (sgnn7) [inline internal/proxyservice/proxy_service.go]: It seems that we are leaking implementation details up to the level where they might not need to be visible - I think the knowledge of how to squash and parse http service configs probably needs to be in some `.... .http ...` package since the `proxyservice` should ideally not need to know anything other than how to start/stop services

File: internal/proxyservice/proxy_service.go [hunk 3]
```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
...
+
+	// Now create proper HTTPServiceConfig objects from our map
+	var httpServiceConfigs []HTTPServiceConfig
+	for listenOn, configs := range httpByListenOn {
+		httpServiceConfig := HTTPServiceConfig{
+			SharedListenOn:    listenOn,
+			SubserviceConfigs: configs,
+		}
+		httpServiceConfigs = append(httpServiceConfigs, httpServiceConfig)
+	}
```
Comment (sgnn7) [inline internal/proxyservice/proxy_service.go]: This is somewhat of a hefty method to parse through in terms of logic - any way to split this up into some smaller functions?

File: internal/proxyservice/proxy_service.go [hunk 4]
```diff
@@ -57,78 +59,246 @@ func (s *proxyServices) Stop() error {
...
+	proxyName := fmt.Sprintf("HTTP Proxy on %s", httpSvcConfig.SharedListenOn)
+	svcLogger := s.logger.CopyWith(proxyName, s.logger.DebugEnabled())
+
+	newSvc, err := httpproxy.NewProxyService(subservices, listener, svcLogger)
+	if err != nil {
+		s.logger.Errorf("could not create http proxy service '%s'", proxyName)
+		return nil, err
+	}
+	return newSvc, nil
+}
```
Comment (sgnn7) [inline internal/proxyservice/proxy_service.go]: I'm guessing that this method's implementation (or http-specific parts) might be a better fit for its own module/package to split concerns

File: internal/proxyservice/proxy_service.go (hunk 2 above)
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: This is exactly right.  You'll note the comment at the top of the method :)

> // TODO: Eventually it should get moved to the config package, and this
> //   transformation should occur at the border.

File: internal/proxyservice/proxy_service.go (hunk 3 above)
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: Definitely.  I'll plan to do this Monday.

File: internal/proxyservice/proxy_service.go (hunk 4 above)
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: I'll think about this some more, but do keep in mind in that:
```go
	newSvc, err := httpproxy.NewProxyService(subservices, listener, svcLogger)
```
which is the punchline of this method is already split out, and that's where the http specific logic will this.  This method you're commenting on here is only responsible for putting the arguments together that get passed down to the real proxy service.  So it's fairly cohesive as is.  That said, it could probably be improved further.

File: internal/proxyservice/proxy_service.go [hunk 5]
```diff
@@ -57,89 +111,145 @@ func (s *proxyServices) Stop() error {
...
-			servicesToStart = append(servicesToStart, tcpSvc)
-			continue
+
+	// TCP Plugins
+	for _, cfg := range configsByType.TCP {
+		// Validation will have already happened
+		tcpSvc, err := s.createTCPService(cfg, tcpPlugins[cfg.Connector])
+		if err != nil {
+			// TODO: Add Fatalf to our logger and use that
+			s.logger.Panicf("unable to create TCP service '%s'", cfg.Name)
```
Comment (doodlesbykumbi) [inline internal/proxyservice/proxy_service.go]: we should append the error to the panic message so the user has additional information as to the cause

File: pkg/secretless/config/v2/config.go [hunk 6]
```diff
@@ -47,3 +51,115 @@ func NewConfig(v2YAML []byte) (*Config, error) {
...
+) ConfigsByType {
+
+	// Get the nil checks out of the way.
+	var rawConfigs []Service
+	for _, cfg := range uncheckedConfigs {
+		if cfg == nil {
+			// Hard-coding log here is okay since nils should never occur and
+			// we won't be unit testing this.  Instead, we'll make the change
+			// in the TODO_ above and this code will be deleted then.
+			log.Fatalln("Nil configuration is not allowed!")
```
Comment (doodlesbykumbi) [inline pkg/secretless/config/v2/config.go]: This is a convenience function called quite deep in the stack. I don't think it should have this catastrophic side effect. I think that should bubble up to a place where it's more obvious what the issue.

File: internal/proxyservice/proxy_service.go [hunk 7]
```diff
@@ -57,89 +111,145 @@ func (s *proxyServices) Stop() error {
...
+func (s *proxyServices) connectorResources(svc v2.Service) connector.Resources {
+	svcLogger := s.loggerFor(svc.Name)
+	return connector.NewResources(svc.ConnectorConfig, svcLogger)
+}
+
+func (s *proxyServices) loggerFor(name string) logapi.Logger {
+	return s.logger.CopyWith(name, s.logger.DebugEnabled())
+}
+
+func (s *proxyServices) credsRetriever(
```
Comment (doodlesbykumbi) [inline internal/proxyservice/proxy_service.go]: proxyServices seems super busy. there's a lot of responsibility here. nothing comes to mind immediately for how to break it up but just wanted to bring that up. i'll be thinking about it ...

File: internal/plugin/manager/manager.go [hunk 8]
```diff
@@ -16,6 +16,8 @@ import (
 	"syscall"
 	"time"
 
+	plugin2 "github.com/cyberark/secretless-broker/internal/plugin"
```
[the remark this answers is not in the document]
Comment (jonahx) [reply internal/plugin/manager/manager.go]: This file will be removed.  This was just a side-effect of some IDE "Move to..." refactoring.  If it's a problem in the code we're keeping let's address that separately.

File: internal/plugin/manager/manager.go [hunk 9]
```diff
@@ -408,7 +410,7 @@ func (manager *Manager) LoadLibraryPlugins(path string, checksumsFile string) er
 		// We override file listing if we did a verification to prevent additions
 		// to plugins between verification and loading the plugins.
 		var err error
-		if files, err = VerifyPluginChecksums(path, checksumsFile); err != nil {
+		if files, err = so.VerifyPluginChecksums(path, checksumsFile); err != nil {
```
[the remark this answers is not in the document]
Comment (jonahx) [reply internal/plugin/manager/manager.go]: Yeah I feel you.  Go does recommend short names though... how about `sharedobj` as compromise?

File: internal/proxyservice/proxy_service.go [hunk 10]
```diff
@@ -1,3 +1,55 @@
+/*
+Package proxyservice takes a Secretless configuration and available plugins
+and constructs the requires ProxyServices that Secretless will run.
+
+To put this in context, the complete high-level flow is:
+
+	1. Parse the `secretless.yml` into individual “service” configs,
+	corresponding to each service entry in the yml.
+
+	2. Identify http services that share a `listenOn`, because we know those
```
[the remark this answers is not in the document]
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: This is something we've discussed before iirc.  My feeling is that it gets handled with first match.  We could log a warning when there are multiple matches.  And then we'd make sure to document that in the public docs as well.

File: internal/proxyservice/proxy_service.go [hunk 11]
```diff
@@ -57,89 +111,145 @@ func (s *proxyServices) Stop() error {
...
-	credsCopy := svc.Credentials
-	credsRetriever := func() (map[string][]byte, error) {
-		return GetSecrets(credsCopy)
+	// TODO: Add validation somewhere about overlapping listenOns
+	// TODO: v2.NetworkAddress is a value type.  It needs to be moved to its
+	//   own package with no deps (stdlib deps are ok).
+	netAddr := v2.NetworkAddress(config.ListenOn)
+	listener, err := net.Listen(netAddr.Network(), netAddr.Address())
+	if err != nil {
+		s.logger.Errorf("could not create listener on: %s", config.ListenOn)
```
[the remark this answers is not in the document]
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: i don't understand this one.   can you clarify?

File: internal/proxyservice/proxy_service.go (hunk 7 above)
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: let's discuss this afternoon

File: pkg/secretless/config/v2/config.go (hunk 6 above)
Comment (jonahx) [reply pkg/secretless/config/v2/config.go]: Ah, so remember per some of my other TODOs, all this stuff will eventually happen at the border.  So it would be fail fast code, exactly where want it, once we make that change.  I didn't want to make that change yet, however, as it's... not hard exactly, but a can of worms in the same way you updating all the application code to use v2 instead of v1 was.  

TLDR: I agree with you assessment, and plan to solve it, but I think solving it later is the right play.

File: pkg/secretless/config/v2/service.go [hunk 12]
```diff
@@ -126,42 +160,48 @@ func NewService(svcName string, svcYAML *serviceYAML) (*Service, error) {
...
+	return a.split()[0]
+}
+
+// Address returns the "address" part of a network address, eg, "127.0.0.1".
+func (a *NetworkAddress) Address() string {
+	return a.split()[0]
+}
+
+func (a *NetworkAddress) split() []string {
+	return strings.Split(string(*a), "://")
```
[the remark this answers is not in the document]
Comment (jonahx) [reply pkg/secretless/config/v2/service.go]: Yeah we could do that.  I'd be fine with it.   It seems like the convention is to use pointer receivers for methods -- that's the only reason I did it -- but there's no reason we have to.  Making it a value causes some extra copying but it's trivial so it doesn't really matter.

File: pkg/secretless/plugin/so/available_plugins_test.go [hunk 13]
```diff
@@ -34,7 +35,7 @@ func getMockPlugins() AvailablePlugins {
 
 func TestPlugins(t *testing.T) {
 	t.Run("HTTPPlugins", func(t *testing.T) {
-		httpPlugins := getMockPlugins().HTTPPlugins()
+		httpPlugins := HTTPPlugins()
```
[the remark this answers is not in the document]
Comment (jonahx) [reply pkg/secretless/plugin/so/available_plugins_test.go]: Let me look into this later, this wasn't code I wrote.  This just got folded into the PR because of ripple changes from the "Move to..." refactoring the IDE did.

File: pkg/secretless/plugin/so/internal_plugins_test.go [hunk 14]
```diff
@@ -1,16 +1,17 @@
-package plugin
+package so
 
 import (
 	"errors"
 	"testing"
 
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin"
```
[the remark this answers is not in the document]
Comment (jonahx) [reply pkg/secretless/plugin/so/internal_plugins_test.go]: #$%! IDE... I thought it was supposed to hanle this automatically.  Anyway, fixed :)

File: internal/proxyservice/proxy_service.go (hunk 7 above)
Comment (jonahx) [reply internal/proxyservice/proxy_service.go]: closed per discussion on phone

File: pkg/secretless/config/v2/service.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonahx) [reply pkg/secretless/config/v2/service.go]: Done.
